src/crypto/crypto_context.cc

```diff
@@ -222,6 +227,245 @@ unsigned long LoadCertsFromFile(  // NOLINT(runtime/int)
   }
 }
 
+enum TrustStatus { UNSPECIFIED, TRUSTED, DISTRUSTED };
+
+char* bioToCString(BIOPointer* bio) {
+  const int len = BIO_pending(bio->get());
+  char* result = reinterpret_cast<char *>(calloc(len + 1, 1));
+  BIO_read(bio->get(), result, len);
+
+  return result;
+}
+
+char* getIssuer(ncrypto::X509View x509_view) {
+  auto bio = x509_view.getIssuer();
+
+  return bioToCString(&bio);
+}
+
+char* getSubject(ncrypto::X509View x509_view) {
+  auto bio = x509_view.getSubject();
+
+  return bioToCString(&bio);
+}
+
+bool IsSelfSigned(X509* cert) {
+  auto subject = X509_get_subject_name(cert);
+  auto issuer = X509_get_issuer_name(cert);
+
+  if (X509_NAME_cmp(subject, issuer) == 0) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+enum TrustStatus IsTrustDictionaryTrustedForPolicy(
+                                       CFDictionaryRef trust_dict
+                                    ) {
+  // Trust settings may be scoped to a single application
+  // skip as this is not supported
+  if (CFDictionaryContainsKey(trust_dict, kSecTrustSettingsApplication)) {
+    return UNSPECIFIED;
+  }
+
+  // Trust settings may be scoped using policy-specific constraints. For
+  // example, SSL trust settings might be scoped to a single hostname, or EAP
+  // settings specific to a particular WiFi network.
+  // As this is not presently supported, skip any policy-specific trust
+  // settings.
+  if (CFDictionaryContainsKey(trust_dict, kSecTrustSettingsPolicyString)) {
+    return UNSPECIFIED;
+  }
+
+  int trust_settings_result = kSecTrustSettingsResultTrustRoot;
+  if (CFDictionaryContainsKey(trust_dict, kSecTrustSettingsResult)) {
+    CFNumberRef trust_settings_result_ref = (CFNumberRef) CFDictionaryGetValue(
+        trust_dict, kSecTrustSettingsResult);
+
+    CFNumberGetValue(trust_settings_result_ref, kCFNumberIntType,
+                     &trust_settings_result);
+
+    if (!trust_settings_result_ref) {
+      return UNSPECIFIED;
+    }
+
+    if (trust_settings_result == kSecTrustSettingsResultDeny) {
+      return DISTRUSTED;
+    }
+    return trust_settings_result == kSecTrustSettingsResultTrustRoot ||
```

Comment:
Isn't this hack only valid for root certificates? https://github.com/chromium/chromium/blob/main/net/cert/internal/trust_store_mac.cc#L175-L180 Otherwise it is wrong to consider kSecTrustSettingsResultTrustRoot on non-root certificates according to https://developer.apple.com/documentation/security/sectrustsettingscopytrustsettings(_:_:_:)

Reply:
The OS will give `kSecTrustSettingsResultTrustAsRoot` then, I don't think there's an issue here but I can add the check.